src/Day20/Solution.hs

```diff
@@ -0,0 +1,129 @@
+module Day20.Solution where
+
+import Advent.Parser
+import Advent.Utils
+import Data.Foldable (find)
+import Data.Function
+import Data.Functor
+import Data.List (transpose)
+import qualified Data.Map.Strict as Map
+import Data.Maybe
+import Data.Sequence hiding (reverse, take)
+import qualified Data.Set as Set
+import Text.Parsec hiding (Empty)
+
+part1 :: String -> String
+part1 = show . product . cornerIds . buildGrid . fromRightOrShowError . parseTiles
+
+part2 :: String -> String
+part2 = head . lines
+
+data Tile = Tile {tId :: Int, tContent :: [[Char]]} deriving (Show, Eq)
+
+parseTiles :: String -> Either ParseError (Seq Tile)
+parseTiles = parse (tilesParser <* eof) ""
+  where
+    tilesParser :: Parsec String () (Seq Tile)
+    tilesParser = fromList <$> tileParser `sepEndBy1` newline
+
+    tileParser :: Parsec String () Tile
+    tileParser = Tile <$> (string "Tile " *> intParser <* char ':' <* newline) <*> tileContentParser
+
+    tileContentParser :: Parsec String () [[Char]]
+    tileContentParser = many1 (oneOf ['.', '#']) `sepEndBy1` newline
+
+tUp :: Tile -> String
+tUp = head . tContent
+
+tRight :: Tile -> String
+tRight = map last . tContent
+
+tDown :: Tile -> String
+tDown = last . tContent
+
+tLeft :: Tile -> String
+tLeft = map head . tContent
+
+type Point = (Int, Int)
+
+type Grid = Map.Map Point Tile
+
+buildGrid :: Seq Tile -> Grid
+buildGrid Empty = Map.empty
+buildGrid (x :<| xs) = go (xs, Map.fromList [((0, 0), x)])
+  where
+    go :: (Seq Tile, Grid) -> Grid
+    go (Empty, grid) = grid
+    go (tile :<| tiles, grid)
+      | tileOrientation & isJust = go (tiles, grid `withTile` tileOrientation)
+      | otherwise = go (tiles |> tile, grid)
```

Comment:
This is the program, everything else noise. 

`tileOrientation ` is a brute force search of the grid. Brute force FTW 🎉 .  If the tile has an orientation that fits, we add it to the grid; otherwise, we recycle it to the end of our list of tiles and hope 🤞  that doesn't leave us running an infinite loop. Fortunately, all the tiles fit!


Conceptually, `tileOrientation` considers every "known neighbor" as a candidate to be placed.  For each candidate, it checks with each of its neighbors and tries each of its possible orientations, looking for a fit.  In other words, this thing runs at  Big 0 n ^ 20 😂 .  Or, as I like to say: this thing runs.